File: go/test/test_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/game-development-sdk"
	"github.com/voxgig-sdk/game-development-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestTestEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Test(nil)
		if ent == nil {
			t.Fatal("expected non-nil TestEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := testBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "test." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GAMEDEVELOPMENT_TEST_TEST_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		testRef01Ent := client.Test(nil)
		testRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "test"}, setup.data), "test_ref01"))
		testRef01Data["project_id"] = setup.idmap["project01"]

		testRef01DataResult, err := testRef01Ent.Create(testRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		testRef01Data = core.ToMapAny(testRef01DataResult)
		if testRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if testRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		testRef01Match := map[string]any{
			"project_id": setup.idmap["project01"],
		}

		testRef01ListResult, err := testRef01Ent.List(testRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		testRef01List, testRef01ListOk := testRef01ListResult.([]any)
		if !testRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", testRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(testRef01List), map[string]any{"id": testRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// LOAD
		testRef01MatchDt0 := map[string]any{
			"id": testRef01Data["id"],
		}
		testRef01DataDt0Loaded, err := testRef01Ent.Load(testRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		testRef01DataDt0LoadResult := core.ToMapAny(testRef01DataDt0Loaded)
		if testRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if testRef01DataDt0LoadResult["id"] != testRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func testBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "test", "TestTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read test test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse test test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"test01", "test02", "test03", "project01", "project02", "project03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GAMEDEVELOPMENT_TEST_TEST_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GAMEDEVELOPMENT_TEST_TEST_ENTID": idmap,
		"GAMEDEVELOPMENT_TEST_LIVE":      "FALSE",
		"GAMEDEVELOPMENT_TEST_EXPLAIN":   "FALSE",
		"GAMEDEVELOPMENT_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["GAMEDEVELOPMENT_TEST_TEST_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GAMEDEVELOPMENT_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["GAMEDEVELOPMENT_APIKEY"],
			},
			extra,
		})
		client = sdk.NewGameDevelopmentSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GAMEDEVELOPMENT_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GAMEDEVELOPMENT_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/test_entity_test.rb
# Test entity test

require "minitest/autorun"
require "json"
require_relative "../GameDevelopment_sdk"
require_relative "runner"

class TestEntityTest < Minitest::Test
  def test_create_instance
    testsdk = GameDevelopmentSDK.test(nil, nil)
    ent = testsdk.Test(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "test" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = GameDevelopmentSDK.test(seed, nil)
    seen = base.Test(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = GameDevelopmentConfig.shared_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = GameDevelopmentSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.Test(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = test_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "test." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set GAME_DEVELOPMENT_TEST_TEST_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    test_ref01_ent = client.Test(nil)
    test_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.test"), "test_ref01"))
    test_ref01_data["project_id"] = setup[:idmap]["project01"]

    test_ref01_data_result = test_ref01_ent.create(test_ref01_data, nil)
    test_ref01_data = Helpers.to_map(test_ref01_data_result.respond_to?(:data_get) ? test_ref01_data_result.data_get : test_ref01_data_result)
    assert !test_ref01_data.nil?
    assert !test_ref01_data["id"].nil?

    # LIST
    test_ref01_match = {
      "project_id" => setup[:idmap]["project01"],
    }

    test_ref01_list_result = test_ref01_ent.list(test_ref01_match, nil)
    assert test_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(test_ref01_list_result),
      { "id" => test_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # LOAD
    test_ref01_match_dt0 = {
      "id" => test_ref01_data["id"],
    }
    test_ref01_data_dt0_loaded = test_ref01_ent.load(test_ref01_match_dt0, nil)
    test_ref01_data_dt0_load_result = Helpers.to_map(test_ref01_data_dt0_loaded.respond_to?(:data_get) ? test_ref01_data_dt0_loaded.data_get : test_ref01_data_dt0_loaded)
    assert !test_ref01_data_dt0_load_result.nil?
    assert_equal test_ref01_data_dt0_load_result["id"], test_ref01_data["id"]

  end
end

def test_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "test", "TestTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = GameDevelopmentSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["test01", "test02", "test03", "project01", "project02", "project03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["GAME_DEVELOPMENT_TEST_TEST_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "GAME_DEVELOPMENT_TEST_TEST_ENTID" => idmap,
    "GAME_DEVELOPMENT_TEST_LIVE" => "FALSE",
    "GAME_DEVELOPMENT_TEST_EXPLAIN" => "FALSE",
    "GAME_DEVELOPMENT_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["GAME_DEVELOPMENT_TEST_TEST_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["GAME_DEVELOPMENT_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["GAME_DEVELOPMENT_APIKEY"],
      },
      extra || {},
    ])
    client = GameDevelopmentSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["GAME_DEVELOPMENT_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["GAME_DEVELOPMENT_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
